### ISSUE TITLE
Issue with Finetunning

### ISSUE THREAD

**Issue by melgor** [facebook/fb.resnet.torch#4]
I was looking the code for fine-tuning the ResNet, exactly:
```lua 
linear.weight:zero() -- only trains last layer to start
linear.bias:zero()
```
This line set all the weights to zeros. But this cause that the weights change in that layer will be always zero (gradient will be 0), isn't it?

I made a small test using such lines with fine-tuning my net and then the net was not learning anything. Without this lines everything was fine. 

Could you explain why you set this lines here?  I think this line are incorrect and should be deleted or replaced (by any weight initialization method).

**Comment by colesbury** [comment]
The gradient w.r.t weight in the final layer is non-zero:

`gradWeight = gradOutput * input`

The gradient w.r.t input is initially zero (i.e. only the last layer is fine tuned), but this changes as the weights get updated:

`gradInput = gradOutput * weight`

This works as long as long as you only zero out weights in the last layer. Have you changed other parts of the model?

**Comment by melgor** [comment]
You are right.
I was using fine-tuning for learning embedding using TripletLoss.  And using such Cost Function we have to make sure that weights area non-zero (because then all embedding will be zeros what will not produce ant gradient).
But in case of CrossEntropyCriterion everything will works like you said. Thanks for clarification.
